src/diff_generate.c

```diff
@@ -989,8 +989,13 @@ static int handle_unmatched_new_item(
 	if (git_index_entry_is_conflict(nitem))
 		delta_type = GIT_DELTA_CONFLICTED;
 
-	/* update delta_type if this item is ignored */
-	else if (git_iterator_current_is_ignored(info->new_iter))
+	/* update delta_type if this item is ignored but
+	 * only if it makes a difference -- ignore checks are expensive
+	 */
+	else if ((!contains_oitem ||
+		  DIFF_FLAG_IS_SET(diff, GIT_DIFF_INCLUDE_IGNORED | GIT_DIFF_RECURSE_IGNORED_DIRS |
```

Comment:
Please double-check this set of flags. I tried to be conservative here and have likely enabled the fast path in fewer cases than I could have.